types/json/path_expr.go

```diff
@@ -165,111 +161,217 @@ func (pe PathExpression) ContainsAnyAsterisk() bool {
 	return pe.flags.containsAnyAsterisk()
 }
 
-// ParseJSONPathExpr parses a JSON path expression. Returns a PathExpression
-// object which can be used in JSON_EXTRACT, JSON_SET and so on.
-func ParseJSONPathExpr(pathExpr string) (pe PathExpression, err error) {
-	peCache.mu.Lock()
-	val, ok := peCache.cache.Get(pathExpressionKey(pathExpr))
-	if ok {
-		peCache.mu.Unlock()
-		return val.(PathExpression), nil
-	}
-	peCache.mu.Unlock()
+type stream struct {
+	pathExpr string
+	pos      int
+}
 
-	defer func() {
-		if err == nil {
-			peCache.mu.Lock()
-			peCache.cache.Put(pathExpressionKey(pathExpr), kvcache.Value(pe))
-			peCache.mu.Unlock()
+func (s *stream) skipWhiteSpace() {
+	for ; s.pos < len(s.pathExpr); s.pos++ {
+		if !unicode.IsSpace(rune(s.pathExpr[s.pos])) {
+			break
 		}
-	}()
-
-	// Find the position of first '$'. If any no-blank characters in
-	// pathExpr[0: dollarIndex), return an ErrInvalidJSONPath error.
-	dollarIndex := strings.Index(pathExpr, "$")
-	if dollarIndex < 0 {
-		err = ErrInvalidJSONPath.GenWithStackByArgs(pathExpr)
-		return
 	}
-	for i := 0; i < dollarIndex; i++ {
-		if !isBlank(rune(pathExpr[i])) {
-			err = ErrInvalidJSONPath.GenWithStackByArgs(pathExpr)
-			return
+}
+
+func (s *stream) read() byte {
+	b := s.pathExpr[s.pos]
+	s.pos++
+	return b
+}
+
+func (s *stream) peek() byte {
+	return s.pathExpr[s.pos]
+}
+
+func (s *stream) skip(i int) {
+	s.pos += i
+}
+
+func (s *stream) exhausted() bool {
+	return s.pos >= len(s.pathExpr)
+}
+
+func (s *stream) incAsLongAs(f func(byte) bool) bool {
+	for ; !s.exhausted(); s.skip(1) {
+		if !f(s.peek()) {
+			return true
 		}
 	}
+	return false
+}
 
-	pathExprSuffix := strings.TrimFunc(pathExpr[dollarIndex+1:], isBlank)
-	indices := jsonPathExprLegRe.FindAllStringIndex(pathExprSuffix, -1)
-	if len(indices) == 0 && len(pathExprSuffix) != 0 {
-		err = ErrInvalidJSONPath.GenWithStackByArgs(pathExpr)
-		return
+func parseJSONPathExpr(pathExpr string) (pe PathExpression, err error) {
+	s := &stream{pathExpr: pathExpr, pos: 0}
+	s.skipWhiteSpace()
+	if s.exhausted() || s.read() != '$' {
+		return PathExpression{}, ErrInvalidJSONPath.GenWithStackByArgs(1)
 	}
+	s.skipWhiteSpace()
 
-	pe.legs = make([]pathLeg, 0, len(indices))
+	pe.legs = make([]pathLeg, 0, 16)
 	pe.flags = pathExpressionFlag(0)
 
-	lastEnd := 0
-	for _, indice := range indices {
-		start, end := indice[0], indice[1]
+	var ok bool
+
+	for !s.exhausted() {
```

Comment:
Maybe `skipWhiteSpace()` at the beginning of the loop, instead of the end such that:

1. L210 can be removed
2. every `parseXXX` does not need to skip white spaces

Reply:
after `skipWhiteSpace` we need to re-check if the string is exhausted, nothing better than now

Reply:
Yeah, I just mean to reduce occurence of `skipWhiteSpace` for cleaner code path. It is just a code style suggestion. I won't insist the change.